tests/src/test/scala/cats/tests/MonadRecInstancesTests.scala

```diff
@@ -0,0 +1,37 @@
+package cats
+package tests
+
+import cats.data.{OptionT, Xor, XorT}
+
+class MonadRecInstancesTests extends CatsSuite {
+  def tailRecMStackSafety[M[_]](implicit M: MonadRec[M], Eq: Eq[M[Int]]): Unit = {
+    val n = 50000
```

Comment:
@adelbertc Is it OK to call that unit test a law?

@ceedubs The best I can think of is testing that `tailRecM` is consistent with the naive (non-stack safe) implementation of `tailRecM` that uses recursive `flatMap`s (and is implemented uniformly for any monad)

```scala
def naiveTailRecM[M[_], A, B](a: A)(f: A => M[A Xor B])(implicit M: Monad[M]): M[B] =
  M.flatMap(f(a))(_ match {
    case Xor.Left(a1) => naiveTailRecM(a1)(f)
    case Xor.Right(b) => M.pure(b)
  })
```

(This is also basically the implementation of `tailRecM` for `Free`, where it is OK, because `Free` is lazy.)

Did you have something else in mind?